File: solutions/zig/binary-search/1/binary_search.zig
pub const SearchError = error{
    ValueAbsent,
    EmptyBuffer,
};

pub fn binarySearch(comptime T: type, target: usize, buffer: ?[]const T) ?usize {
    const buff = buffer orelse return null;
    
    if (buff.len == 0) {
        return null;
    }
    
    var low: usize = 0;
    var high: usize = buff.len;
    
    while (low < high) {
        const mid = low + ((high - low) / 2);
        if (buff[mid] == target) {
            return mid;
        } else if (buff[mid] > target) {
            high = mid;
        } else {
            low = mid + 1;
        }
    }
    return null;
}

File: solutions/fortran/binary-search/1/binary_search.f90
! For my Shree DR.MDD
module binary_search
   implicit none

   private

   public find
contains

   function find(array_seq, target) result(position)
      integer, dimension(:), intent(in) :: array_seq
      integer, intent(in) :: target
      integer :: position
      integer :: low
      integer :: high
      integer :: mid

      low = 1
      high = size(array_seq)
      if (high == 0) then
         position = -1
         return
      end if
      if (array_seq(low) > target .or. array_seq(high) < target) then
         position = -1
         return
      end if

      mid = (low + high)/2
      do while (.not. array_seq(mid) == target)
         if (high - low > 1) then
            mid = (high + low)/2
         else
            if (array_seq(low) == target) then
               mid = low
            else
               if (array_seq(high) == target) then
                  mid = high
               else
                  mid = -1
               end if
            end if
            exit
         end if
         if (array_seq(mid) > target) then
            high = mid
         else
            low = mid
         end if
      end do

      position = mid
   end function find

end module binary_search
